core-libs/setup/ssr/engine/cx-common-engine-wrapper.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { StaticProvider } from '@angular/core';
+import { CommonEngine } from '@angular/ssr';
+import { CxServerErrorResponse } from '../error-handling/server-error-response';
+import { PROPAGATE_SERVER_ERROR_RESPONSE } from '../error-handling/server-error-response/propagate-server-error-response';
+import { RenderOptions } from './ng-express-engine';
+
+/**
+ * Wraps the CommonEngine render method to handle propagated server responses caught during server-side rendering.
+ */
+
+export function cxCommonEngineWrapper(
+  engine: CommonEngine,
+  renderOptions: RenderOptions,
+  callback: (err?: Error | null, html?: string) => void
+) {
+  const propagatedServerErrorResponses: (CxServerErrorResponse | string)[] = [];
+
+  const propagateServerErrorResponseProvider = {
+    provide: PROPAGATE_SERVER_ERROR_RESPONSE,
+    useFactory: () => {
+      return (serverErrorResponse: CxServerErrorResponse) => {
+        propagatedServerErrorResponses.push(serverErrorResponse);
+      };
+    },
+  };
+
+  return engine
+    .render({
+      ...renderOptions,
+      providers: [
+        propagateServerErrorResponseProvider,
+        ...(renderOptions.providers ?? []),
+      ] as StaticProvider[],
+    })
+    .then((html) => propagatedServerErrorResponses.push(html))
+    .catch((err) => propagatedServerErrorResponses.push(err))
+    .finally(async () => {
+      const serverResponse = propagatedServerErrorResponses[0];
+      // If the first event is a CxServerErrorResponse, we call the callback with the error.
+      // which will be handled by the error handler in the ExpressJS server.
+      // Otherwise, we call the callback with the rendered html.
+      if (serverResponse instanceof CxServerErrorResponse) {
+        callback(serverResponse, undefined);
+      } else {
+        callback(undefined, serverResponse);
+      }
```

Comment:
nitpick: as far as I understand, the logic is incorrect in case of Errors from APP_INITAILIZERs. Any Error from APP_INITIALIZER is not an instance of CxServerErrorResponse, therefore in line 51 such an Error will be populated via callback as a success parameter. 

Please let's add an unit test for such a case.

Anyways, in the comment above I proposed a simplified approach (without arrays) which by the way gets rid of this incorrect behavior